pkg/controller/job/job_controller.go

```diff
@@ -845,10 +850,17 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 			}
 			jobCtx.podsWithDelayedDeletionPerIndex = getPodsWithDelayedDeletionPerIndex(logger, jobCtx)
 		}
+		if canApplySuccessPolicy(jobCtx) {
+			if msg, met := metSuccessPolicy(logger, job.Spec.SuccessPolicy, *job.Spec.Completions, jobCtx.succeededIndexes); met {
+				jobCtx.finishedCondition = newCondition(batch.JobSuccessCriteriaMet, v1.ConditionTrue, batch.JobReasonSuccessPolicy, msg, jm.clock.Now())
+			}
+		}
 	}
+
 	suspendCondChanged := false
 	// Remove active pods if Job failed.
-	if jobCtx.finishedCondition != nil {
+	if finCond := jobCtx.finishedCondition; finCond != nil &&
+		!(feature.DefaultFeatureGate.Enabled(features.JobSuccessPolicy) && finCond.Type == batch.JobComplete) {
```

Comment:
We may have a helper function, say `isFailure(jobCtx.finishedCondition)` which seems reusable below.

Reply:
It makes sense.

Reply:
I removed these checks based on https://github.com/kubernetes/kubernetes/pull/123412#discussion_r1502540899